File: challenges/binarySearch.js
//-----------------------------------------------//
// Search the array of prime numbers for a given value using a binary search
//-----------------------------------------------//

const primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97];


function doSearch(array, targetValue) {
  // looking for the index of our target value
  // min is starting point
  // max is the length of array or total number of induces
  // guess is current index number in the array
  // array[guess] is the index value
  let min = 0;
  let max = array.length - 1;
  let guess;
  while (min <= max) {
    // while we are still looping through the array...or while we havnt reached the max or end of array
    // guess equals the closest integer rounded down, find the index
    // guess starts as the halfway point index of the array
    guess = Math.floor((max + min) / 2);
    // if it finds the index, return true
    // else look to the left or right accordingly
    if (array[guess] === targetValue) {
      return guess;
    } else if (array[guess] < targetValue) {
      // if the starting index is less than our target value, than we dont need all the numbers below it.
      // new min is current array[guess] or (the index number) 1 greater than current value, and look to the right.
      min = guess + 1;
    } else {
      // else if the targetVal is greater than the current index, disregard all the numbers to the right and new index is one to down to the left .
      max = guess - 1;
    }
  }
  // if it doesn't find it return false.
  return -1;
}
console.log("doSearch", doSearch(primes, 73));



// Using Recursion
function binaryRecursion(arr, targetValue, min, max) {
  if (min > max) return false;
  let middleIndex = Math.floor((max + min) / 2);
  if (arr[middleIndex] === targetValue) return middleIndex;
  if (arr[middleIndex] > targetValue) {
    return binaryRecursion(arr, targetValue, min, middleIndex - 1);
  } else {
    return binaryRecursion(arr, targetValue, middleIndex + 1, max);
  }
}

console.log("doSearchRecursion", binaryRecursion(primes, 73, 0, primes.length - 1));

File: challenges/python/binary_search.py
#----------------------------------------------#
# Search the List of prime numbers for a given value using a binary search
#-----------------------------------------------#
import math

primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]

class DoBinarySearch(object):
  def do_search(self, input_list, target):
    min = 0
    max = len(input_list) - 1
    guess = 0
    while min <= max:
      guess = math.floor((min + max) // 2)
      if input_list[guess] == target:
        return guess
      elif input_list[guess] < target:
        min = guess + 1
      elif input_list[guess] > target:
        max = guess - 1
    return -1

solution_1 = DoBinarySearch()
print(solution_1.do_search(primes, 73)) # 20
print("--------------------------")




class DoBinarySearchRecursion(object):
  def do_search_recursion(self,input_list, target, min, max):
    #min = 0
    #max = len(input_list) - 1
    if min > max:
      return -1
    # finding middle of list by getting average.
    middle = math.floor((min + max) // 2)

   # if it equals, it is successful so return it.
    if input_list[middle] == target:
      return middle

    # if middle is greater than target, we don't need everything above that number
    # so set new max to middle -1 (b/c doesn't equal current number either.)
    if input_list[middle] > target:
      return self.do_search_recursion(input_list, target, min, (middle - 1))

    # if middle less than target, we don't need anything below that number.
    # so set new min to middle + 1 (b/c doesn't equal current number either.)
    if input_list[middle] < target:
      return self.do_search_recursion(input_list, target, (middle + 1), max)


solution_2 = DoBinarySearchRecursion()
print(solution_2.do_search_recursion(primes, 73, 0, len(primes) - 1)) # 20
print("--------------------------")



# Do without passing in a min and max
# (WIP - todo)
def do_search_recursion_2(input_list, target):
  min = 0
  max = len(input_list) - 1
  if min > max:
    return -1
  middle = math.floor((min + max) // 2)
  if input_list[middle] == target:
    return middle
  if input_list[middle] > target:
    new_input_list = range(input_list[min], input_list[middle - 1])
    return do_search_recursion_2(new_input_list, target)
  if input_list[middle] < target:
    new_input_list = range(input_list[middle + 1], input_list[max])
    return do_search_recursion_2(new_input_list, target)

# print("recursion", do_search_recursion_2(primes, 73))
print("--------------------------")
